Update to Python 3.8 minimum.

bjrichardwebster opened the pull request.

Purg commented on docs/release_notes/pending_release.rst:
```diff
@@ -4,10 +4,15 @@ Pending Release Notes
 
 * Updated CI unittests workflow to include codecov reporting.
   Reduced CodeCov report submission by skipping this step on scheduled runs.
+* Updated CI unittests to reflect new minimum support `python = "^3.8"`.
```
Minor: sphinx/RST is going to want a newline between bullets.

Purg commented on pyproject.toml:
```diff
@@ -18,39 +18,40 @@ classifiers = [
-pillow = "^8.3.2"
-smqtk-core = ">=0.18.0"
+python = "^3.8"
+pillow = ">=9.2.0"
+numpy = ">=1.22"
```
Are these updated version lower-bounds required? I.e. could we instead have `>=` inequalities with the previous versions?

bjrichardwebster replied:
Pillow:
There were numerous dependabot alerts, this version is the lowest to address them all.

Numpy:
Dependabot alert.

Neither are transitory packages.

Purg commented on pyproject.toml:
```diff
@@ -18,39 +18,40 @@ classifiers = [
+mock = ">=4.0.3"
+pytest = ">=7.2.0"
+pytest-cov = ">=2.11.1"
+requests = ">=2.31.0"
```
Looks like a new dependency. Is this required due to some ripple effect?

bjrichardwebster replied:
The change from ^ to >= is to allow versions for python3.11, the requests import is a mistake I'll remove.

Purg reviewed (commented):
Were you thinking you were going to address the new mypy typechecking issues with this branch?

bjrichardwebster commented:
When addressing the mypy type checking errors I came across three major themes:

1) numpy added a couple different explicit typing features in 1.20, 1.21, and 1.22. This caused a few cascading type checking issues that I've addressed by Union between the numpy typing definition with the old Sequence definition which handles non-numpy inputs

2) A couple of test files were attempting to override an abstract function with Optional parameter with a non-Optional function. This is no longer allowed because mypy has changed its default to no_implicit_optional=True. See [incompatible-overrides](https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides).

3) ImageReader.load_as_matrix has an Optional return while numpy.testing.assert_allclose parameters do not allow Optional. Since all mypy errors connected to this difference were in the unit tests, I added the `type: ignore` to those lines. In the unit tests we know None will never be returned in those situations and we want to keep the unit test focused, so adding an extra explicit assertion for None (instead of type:ignore) would add an assertion outside the scope of the particular unit test.
